File: BeachRehberi.API/BeachRehberi.API/Models/PagedResponse.cs
namespace BeachRehberi.API.Models;

public class PagedResponse<T>
{
    public List<T> Items { get; set; } = new();
    public int TotalCount { get; set; }
    public int TotalPages { get; set; }
    public int CurrentPage { get; set; }
    public int PageSize { get; set; }

    public PagedResponse(List<T> items, int count, int page, int pageSize)
    {
        Items = items;
        TotalCount = count;
        CurrentPage = page;
        PageSize = pageSize;
        TotalPages = (int)Math.Ceiling(count / (double)pageSize);
    }
}

File: mobile/lib/core/network/paged_response.dart
class PagedResponse<T> {
  const PagedResponse({
    required this.items,
    required this.page,
    required this.pageSize,
    required this.totalCount,
  });

  final List<T> items;
  final int page;
  final int pageSize;
  final int totalCount;

  factory PagedResponse.fromJson(
    Map<String, dynamic> json,
    T Function(Map<String, dynamic> itemJson) itemParser,
  ) {
    final rawItems = json['items'];
    final items = rawItems is List
        ? rawItems
            .whereType<Map<String, dynamic>>()
            .map(itemParser)
            .toList(growable: false)
        : <T>[];

    return PagedResponse<T>(
      items: items,
      page: _asInt(json['page']) ?? _asInt(json['currentPage']) ?? 1,
      pageSize: _asInt(json['pageSize']) ?? items.length,
      totalCount: _asInt(json['totalCount']) ?? items.length,
    );
  }

  static int? _asInt(Object? value) {
    if (value is int) return value;
    if (value is num) return value.toInt();
    return int.tryParse(value?.toString() ?? '');
  }
}
